File: node/nodes/output_node.py
import os
from typing import Dict, Any

from metadata import Metadata
from node import Node, register_node_class


@register_node_class
class OutputNode(Node):

    node_type = "Output"

    def get_input_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "files": {
                "type": "FILEPATH_ARRAY",
                "required": True,
                "widget": "LABEL",
                "default": set()
            },
            "prefix": {
                "type": "STRING",
                "required": False,
                "widget": "INPUT",
                "default": ""
            },
            "folder": {
                "type": "STRING",
                "required": False,
                "widget": "INPUT",
                "default": ""
            }
        }

    def get_output_types(self) -> Dict[str, Dict[str, Any]]:
        return {}

    def get_display_info(self) -> Dict[str, str]:
        return {
            "category": "IO",
            "description": "Output files to a folder",
            "color": "#AED581",
            "icon": "OutputIcon"
        }

    def exec(self, files, prefix, folder, meta: Metadata) -> Dict[str, Any]:
        for filepath in files:
            from storage_client import StorageClient
            StorageClient.store_from_workspace_to_scene(
                scene_id=meta.scene_id,
                source=filepath,
                folder=folder if folder else None,
                prefix=prefix if prefix else None
            )

        return {}

    def validate(self, files, prefix, folder, meta) -> None:
        pass

File: engine/src/main/java/com/example/mypixel/model/node/OutputNode.java
package com.example.mypixel.model.node;

import com.example.mypixel.model.Parameter;
import com.example.mypixel.model.ParameterType;
import com.example.mypixel.model.ParamsMap;
import com.example.mypixel.model.Widget;
import com.example.mypixel.service.FileHelper;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.lang.NonNull;

import java.util.HashSet;
import java.util.Map;


@MyPixelNode("Output")
public class OutputNode extends Node {

    @JsonCreator
    public OutputNode(
            @JsonProperty("id") @NonNull Long id,
            @JsonProperty("type") @NonNull String type,
            @JsonProperty("inputs") Map<String, Object> inputs) {
        super(id, type, inputs);
    }

    @Override
    public Map<String, Parameter> getInputTypes() {
        return ParamsMap.of(
                "files", Parameter.builder()
                        .type(ParameterType.FILEPATH_ARRAY)
                        .required(true)
                        .widget(Widget.LABEL)
                        .build(),
                "prefix", Parameter.builder()
                        .type(ParameterType.STRING)
                        .required(false)
                        .widget(Widget.INPUT)
                        .build(),
                "folder", Parameter.builder()
                        .type(ParameterType.STRING)
                        .required(false)
                        .widget(Widget.INPUT)
                        .build()
        );
    }

    @Override
    public Map<String, Object> getDefaultInputs() {
        return ParamsMap.of(
                "files", new HashSet<>(),
                "prefix", "",
                "folder", ""
        );
    }

    @Override
    public Map<String, Parameter> getOutputTypes() {
        return ParamsMap.of();
    }

    @Override
    public Map<String, String> getDisplayInfo() {
        return ParamsMap.of(
                "category", "IO",
                "description", "Output files to a folder",
                "color", "#AED581",
                "icon", "OutputIcon"
        );
    }

    @SuppressWarnings("unchecked")
    @Override
    public Map<String, Object> exec() {
        HashSet<String> files = (HashSet<String>) inputs.get("files");
        String prefix = (String) inputs.getOrDefault("prefix", null);
        String folder = (String) inputs.getOrDefault("folder", null);

        Map<String, Object> outputs = Map.of();

        for (String filepath : files) {
            FileHelper.storeFromTaskToSceneContext(sceneId, filepath, folder, prefix);
        }

        return outputs;
    }

    @Override
    public void validate() {

    }
}
